### ISSUE TITLE
Schemas with shared members 

### ISSUE THREAD

**Issue by rmosolgo** [rmosolgo/graphql-ruby#269]
There's a compelling case for multiple schemas built from the same set of schema members (eg, instances of `GraphQL::ObjectType` or `GraphQL::Field`): 

- Schema A is private and exposes all kinds of data; it's for internal, server-to-server communication 
- Schema B is public and exposes a subset of Schema A; it's for external users (or a JavaScript front-end). 

In this case, some members from Schema A should not even be _visible_ in Schema B: they shouldn't appear in introspection queries and attempts to access them should be treated as accessing a non-existing member. 

Since Schema B is a subset of Schema A, it's desirable to share code between the two. However, a simple problem makes this hard right now: schema members are held in common between the two schemas, so any attempt to modify _one_ of those objects will affect _both_ schemas. 

How can we support this use case? I can think of a few possibilities (and more than a few problems), here they are: 

- __Do nothing.__ Users should write their own Ruby code to describe their system, then build two independent schemas composed of new objects for each schema. This is basically how programmatically-generated fields work currently: you have to regenerate the field each time you use it. Except now we're talking about an entire schema! __Pros__ are: very easy. __Cons__ are: we waste a lot of user time for problem that could be common. Besides that, we essentially ask people to reimplement GraphQL schema definition, which is sad. 
- __Schema deep-copy.__ Provide an API like `Schema#filter(&block)` which recursively copies Schema members, calling `block` to determine whether each member should be included in the new schema. This could work, but we'd probably still end up sharing some user-provided values, like resolve functions or argument default values. (Or, could we copy those too?) __Pros__ are: it would work. __Cons__ are: it will be hard to get right. Depending on what we copy, and how we copy it, it could lead to some surprising behaviors if users modify objects after-the-fact. 
- __Dynamic schema member resolution.__ Accept query-specific input (eg `context[:permission]`) and check that anytime you fetch schema members (types, fields, arguments). __Pros__ are: this is easy for users. You hook up metadata to your objects, then define how that query-specific input should be checked against metadata. __Cons__ are: it will be hard to find all of these places in the gem and update them. Until now, we have assumed that a schema's members are fixed after `Schema.define`. If we introduce query-level schema membership, it will be really hard to find where those assumptions are made and ... unmake them!
- Ok, here's a weird one: __Schema "masks".__ Assuming Schema B is a subset of Schema A, what if we could use Schema B as a "mask" for Schema A, meaning that: 
  - Introspection queries are resolved against Schema B 
  - Queries are validated against Schema B 
  - Queries are executed against Schema A 

  The API could be `SchemaC = Schema::Mask.new(schema: SchemaA, mask: SchemaB)`. Then, queries to `SchemaC` would be handled as described above. `SchemaB` could be derived by stringifying & reloading `SchemaA`, then removing some parts (or, remove parts while stringifying, whatever). __Pros__ are: this is relatively easy to implement. Only introspection and validation needs to be changed, so that `Schema::Mask` can quack like a Schema and handle those behaviors differently. Schema A's assumption of immutability can remain the same.  __Cons__ are: it's a weird idea. (Anything else bad about this?) 

As you can tell by my analysis, I think `Schema::Mask` is so crazy, it just might work! 

Are there other ways to support a setup like this, or any pros or cons that I've overlooked for the implementations described above?

/cc @josh @brandonblack

**Comment by brandonblack** [comment]
@rmosolgo excellent write-up. Thank you for kicking this off.

How do you imagine defining schema objects would work under the `Schema::Mask` approach? Meaning, how would we define what schema A and B actually contain? 

The issue right now is primarily that the current approach to defining schema objects shares a lot at class levels. This is why a deep copy (which is effectively what we've done in our own hack job internally) is a solution that works with the exception of [a few places where graphql-ruby is doing an exact object comparison](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/static_validation/rules/fragment_spreads_are_possible.rb#L35).

I actually dig the idea of a mask, but I feel like how we define schema objects would still need to change to make it possible.

**Comment by rmosolgo** [comment]
Here's the API I imagine: 

```ruby
PrivateSchema = Schema.define do 
  # Make the big schema here 
  # It includes private members 
end 

# Dump the schema to GraphQL schema language -- but only dump a subset of members
# (The API with a block doesn't exist yet -- but Schema::Printer does)
public_schema_string = Schema::Printer.print_schema(PrivateSchema) { |obj| !obj.metadata[:private] } 

# Make a new schema instance -- but this one has no resolvers. 
# The good news is, we've made all kinds of new objects here, nothing is shared with PrivateSchema 
intermediate_public_schema = Schema::Loader.load(public_schema_string) 

# Now, make a Schema::Mask which has 
# - intermediate_public_schema for responding to introspection queries 
# - intermediate_public_schema for validating incoming documents 
# - PrivateSchema for resolving fields 
PublicSchema = Schema::Mask.new(schema: PrivateSchema, mask: intermediate_public_schema) 


# PrivateSchema responds to private stuff: 
PrivateSchema.execute("{ viewer { name, password } }") 
# => { "data" => { "viewer" => { "name" => "rmosolgo", "password" => "p4ssw0rd" } } } 

# But PublicSchema only responds to public stuff: 
PublicSchema.execute("{ viewer { name, password } }")
# => { "errors" => [ { "message" => "No field 'password' on type 'User'", ... } ] } 
PublicSchema.execute("{ viewer { name } }")
# => { "data" => { "viewer" => { "name" => "rmosolgo" } } } 
```

In fact, a lot of these operations could be wrapped into one nice method: 

```ruby 
class Schema 
  # Create a schema like `self`, but which pretends to only have  
  # members for which `block` returns truthy 
  # @return [Schema::Mask] A schema-like object which uses a superset schema for resolving fields
  def create_mask(&block) 
    filtered_string = Schema::Printer.print_schema(self, &block) 
    mask_schema = Schema::Loader.load(filtered_string) 
    Schema::Mask.new(schema: self, mask: mask_schema) 
  end 
end 
```

Then, the user code is crazy simple: 

```ruby 
PublicSchema = PrivateSchema.create_mask { |obj| MyApp.private?(obj) } 
```

Ironically, as I type, I realize we're one method name away from _pretending_ to deep-copy! 

```ruby 
# Doesn't actually filter, but creates a mask over `PrivateSchema`
PublicSchema = PrivateSchema.filter { |obj| MyApp.private?(obj) } 
```

Implementation-wise: 

- Filtered schema printing [almost exists already](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema/printer.rb#L29-L34), but we have to extend it somehow to also filter `GraphQL::Field`s, `GraphQL::Argument`s, `GraphQL::EnumValue`s ... anything else? Ok, it will actually be kind of tricky, because you also have to check field types, argument types, for example, don't include a field if its return type would be hidden.
- We need a little "misdirection point", like `Schema#introspection_target`, for which `Schema` returns `self`, but `Schema::Mask` returns `@mask`. 
- Then, use `instrospection_target` for [building introspection fields](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema.rb#L142-L147) (The introspection field APIs would need updating, too)
- Also, use `introspection_target` for [validating incoming queries](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/query.rb#L161)

That's all I can think of right now!!

**Comment by rmosolgo** [comment]
I started looking for places where fields are _actually_ looked up in the gem right now. After 8627140, I think they're _all_ hitting `Schema#get_field`, which is good.

**Comment by rmosolgo** [comment]
Started taking a stab here: https://github.com/rmosolgo/graphql-ruby/compare/masks

That's actually a dynamic implementation, but at schema-level. I think if we can get _that_ to work, then we could move `visible_x?` methods to `Query`, and get a really nice visibility API, where you can specify visibility at schema-level (eg public vs private), and then _again_ at query-level (eg feature flags).

**Comment by rmosolgo** [comment]
It's a bit of a pivot but as I started, know, _actually_ writing code, I realized that filtering everything is going to be really hard anyways, so why try to get the most out of it? 

Maybe the runtime costs of dynamic filtering will be high, but I think if that's the case, it's better to offset the cost by cleaning up waste elsewhere (I'm sure there is some ... 😬 ).

**Comment by rmosolgo** [comment]
To figure out: how to handle an argument whose default value is hidden? Hide the argument? :boom: ?

**Comment by rmosolgo** [comment]
Merged #300 which provides an API for _hiding_ some schema members on a per-query basis, effectively generating a dynamic subset of the schema

(Or, if someone wants to go the copy-based route, #357 adds `#redefine` to schema members to create shallow copies)
